Sources/TensorFlow/Operators/Basic.swift

```diff
@@ -148,6 +148,18 @@ public extension Tensor {
         reshaped(toShape: other.shapeTensor)
     }
 
+    /// Helper function for reduction ops.
+    /// Returns 1D Tensor, the output shape as if keepDims were set to True.
+    func reducedShape(axes: Tensor<Int32>) -> TensorShape {
```

Comment:
It seems like `reducedShape` is ported from [`def reduced_shape(input_shape, axes)`](https://github.com/tensorflow/tensorflow/blob/8023e0736756b93308cbe847e5e6afbfb625f8d8/tensorflow/python/ops/math_ops.py#L3490).

- Since `reducedShape` is a helper function, please make it private rather than public.
- Since `reducedShape` is an operation on tensor shapes, please change it to take `TensorShape` directly instead of `Tensor`.
- Since `axes` is always 1-D, please change its type to `[Int32]`. (It should be more efficient to perform shape operations on non-`Tensor` types, like `[Int32]` and `Tensor`, because `Tensor` operations have more overhead.)

How about redefining `reducedShape` as a private method on `TensorShape`:
```suggestion
private extension TensorShape {
  /// Returns the reduced shape along the given axes. For each axis, the dimension is set to 1.
  ///
  /// Adapted from `reduced_shape` in Python TensorFlow:
  /// https://github.com/tensorflow/tensorflow/blob/8023e0736756b93308cbe847e5e6afbfb625f8d8/tensorflow/python/ops/math_ops.py#L3490
  func reducedShape(alongAxes axes: [Int32]) -> TensorShape { ... }
}
```

Reply:
Sure, just a little problem over here:

If I define it as `private`, I get `error: 'reducedShape' is inaccessible due to 'fileprivate' protection level`.

If I try it with `internal`, I get error: `instance method 'reducedShape(axes:)' is internal and cannot be referenced from an '@inlinable' function`.

I added reducedShape as public extension in `Sources/TensorFlow/Core/TensorShape.swift` for now. Can you help me with the `private` problem? Thanks.

Reply:
Sorry for the thrash! Let's actually inline `reducedShape`: https://github.com/tensorflow/swift-apis/pull/550#discussion_r344675367.